File: tests/test_p4_candidate_features.py
"""P4 candidate-feature extractor tests.

Verifies the offline Python extractor:
  - builds a versioned mlb-candidate-features-v1.0 vector from stored snapshots
  - enforces promotion-safety (post-pitch evidence -> null, temporal_rejection)
  - emits null + missing_<feature> for absent/insufficient families
  - never embeds outcomes
  - coverage summary is accurate
"""

from __future__ import annotations

import json
import os
import sqlite3
import tempfile
from datetime import datetime, timedelta, timezone

import pytest

from src.candidate_features import (
    build_candidate_dataset,
    build_candidate_vector,
    candidate_coverage_summary,
    CANDIDATE_FEATURE_SCHEMA_VERSION,
    CANDIDATE_FEATURE_NAMES,
)


def _make_db() -> str:
    fd, db_path = tempfile.mkstemp(suffix=".sqlite")
    os.close(fd)
    conn = sqlite3.connect(db_path)
    conn.execute(
        "CREATE TABLE feature_snapshots ("
        "game_pk TEXT NOT NULL, feature_group TEXT NOT NULL, date_ymd TEXT NOT NULL, "
        "payload TEXT NOT NULL, timestamp TEXT NOT NULL, "
        "PRIMARY KEY (game_pk, feature_group))"
    )
    conn.commit()
    conn.close()
    return db_path


def _write_snapshot(conn, game_pk, group, date_ymd, payload, timestamp=None):
    conn.execute(
        "INSERT OR REPLACE INTO feature_snapshots (game_pk, feature_group, date_ymd, payload, timestamp) "
        "VALUES (?, ?, ?, ?, ?)",
        (game_pk, group, date_ymd, json.dumps(payload), timestamp or "2026-07-21T17:00:00Z"),
    )
    conn.commit()


def _statcast_payload(as_of="2026-07-21T17:00:00Z"):
    return {
        "as_of_utc": as_of,
        "fetched_at_utc": as_of,
        "away": {"xwoba": 0.31, "xslg": 0.39, "barrel_rate": 0.07, "hard_hit_rate": 0.38, "sample_pa": 120},
        "home": {"xwoba": 0.33, "xslg": 0.42, "barrel_rate": 0.09, "hard_hit_rate": 0.41, "sample_pa": 118},
    }


def _arsenal_payload(as_of="2026-07-21T17:00:00Z"):
    return {
        "as_of_utc": as_of,
        "away": {"overall_stuff_plus": 105, "stuff_vs_lhh": 103, "stuff_vs_rhh": 107, "sample_pitches": 60},
        "home": {"overall_stuff_plus": 98, "stuff_vs_lhh": 96, "stuff_vs_rhh": 100, "sample_pitches": 55},
    }


def _bvp_payload(as_of="2026-07-21T17:00:00Z"):
    return {
        "as_of_utc": as_of,
        "away": {"ops": 0.75, "plate_appearances": 60},
        "home": {"ops": 0.68, "plate_appearances": 55},
    }


# ---------- Schema + hash ----------

def test_candidate_schema_version_is_separate_from_control():
    assert CANDIDATE_FEATURE_SCHEMA_VERSION == "mlb-candidate-features-v1.0"
    assert "candidate" in CANDIDATE_FEATURE_SCHEMA_VERSION


def test_candidate_feature_names_covers_all_seven_families():
    names = set(CANDIDATE_FEATURE_NAMES)
    # spot-check one feature per family
    assert "homeTeamXwoba" in names  # statcast
    assert "awayStarterStuffPlus" in names  # arsenal
    assert "homeStarterExpectedInnings" in names  # expected innings
    assert "awayBullpenQuality" in names  # bullpen
    assert "homeLineupWeightedOps" in names  # lineup
    assert "awayLineupBvpOps" in names  # bvp
    assert "homeStarterXera" in names  # xera


# ---------- Extraction from stored snapshots ----------

def test_extract_builds_vector_from_promotion_safe_snapshots():
    db_path = _make_db()
    try:
        conn = sqlite3.connect(db_path)
        _write_snapshot(conn, "7001", "statcast_team", "2026-07-21", _statcast_payload())
        _write_snapshot(conn, "7001", "pitch_arsenal", "2026-07-21", _arsenal_payload())
        _write_snapshot(conn, "7001", "bvp", "2026-07-21", _bvp_payload())
        conn.close()

        rows = build_candidate_dataset(
            db_path,
            game_rows=[{"game_pk": "7001", "date_ymd": "2026-07-21",
                        "as_of_utc": "2026-07-21T17:00:00Z",
                        "first_pitch_utc": "2026-07-21T23:05:00Z"}],
        )
        assert "7001" in rows
        row = rows["7001"]
        assert row.schema_version == CANDIDATE_FEATURE_SCHEMA_VERSION
        v = row.feature_vector
        assert v["awayTeamXwoba"] == 0.31
        assert v["homeTeamXwoba"] == 0.33
        assert v["awayStarterStuffPlus"] == 105
        assert v["homeStarterStuffPlus"] == 98
        assert v["awayLineupBvpOps"] == 0.75
        # families without snapshots -> null + missing flag
        assert v["awayStarterExpectedInnings"] is None
        assert v["missing_awayStarterExpectedInnings"] == 1
        assert v["missing_awayTeamXwoba"] == 0
        # hash is deterministic
        assert row.feature_vector_hash == v["featureVectorHash"]
        assert len(row.feature_vector_hash) == 64
        # provenance
        assert set(["statcast_team", "pitch_arsenal", "bvp"]).issubset(set(row.families_available))
        assert "expected_innings" in row.families_missing
    finally:
        os.unlink(db_path)


def test_extract_never_embeds_outcomes():
    """Candidate vector must carry no outcome/winner/label field."""
    db_path = _make_db()
    try:
        conn = sqlite3.connect(db_path)
        _write_snapshot(conn, "7002", "statcast_team", "2026-07-21", _statcast_payload())
        conn.close()
        rows = build_candidate_dataset(
            db_path,
            game_rows=[{"game_pk": "7002", "date_ymd": "2026-07-21",
                        "as_of_utc": "2026-07-21T17:00:00Z",
                        "first_pitch_utc": "2026-07-21T23:05:00Z"}],
        )
        v = rows["7002"].feature_vector
        for key in v:
            kl = key.lower()
            assert "won" not in kl
            assert "winner" not in kl
            assert "outcome" not in kl
            assert "score" not in kl
            assert "result" not in kl
    finally:
        os.unlink(db_path)


# ---------- Temporal safety ----------

def test_extract_rejects_post_pitch_evidence():
    """A payload whose as_of is post-pitch must be rejected for that family."""
    db_path = _make_db()
    try:
        conn = sqlite3.connect(db_path)
        # post-pitch as_of
        _write_snapshot(conn, "7003", "statcast_team", "2026-07-21",
                        _statcast_payload(as_of="2026-07-22T00:30:00Z"))
        conn.close()
        rows = build_candidate_dataset(
            db_path,
            game_rows=[{"game_pk": "7003", "date_ymd": "2026-07-21",
                        "as_of_utc": "2026-07-22T00:30:00Z",
                        "first_pitch_utc": "2026-07-21T23:05:00Z"}],
        )
        row = rows["7003"]
        assert "statcast_team" not in row.families_available
        assert "statcast_team" in row.families_missing
        assert any("statcast_team:post_pitch_source" in t for t in row.temporal_rejections)
        assert row.feature_vector["awayTeamXwoba"] is None
        assert row.feature_vector["missing_awayTeamXwoba"] == 1
    finally:
        os.unlink(db_path)


def test_extract_rejects_missing_first_pitch():
    """No first pitch known -> cannot prove safety -> family dropped."""
    db_path = _make_db()
    try:
        conn = sqlite3.connect(db_path)
        _write_snapshot(conn, "7004", "statcast_team", "2026-07-21", _statcast_payload())
        conn.close()
        rows = build_candidate_dataset(
            db_path,
            game_rows=[{"game_pk": "7004", "date_ymd": "2026-07-21",
                        "as_of_utc": "2026-07-21T17:00:00Z",
                        "first_pitch_utc": None}],
        )
        row = rows["7004"]
        assert "statcast_team" in row.families_missing
        assert any("missing_first_pitch" in t for t in row.temporal_rejections)
    finally:
        os.unlink(db_path)


# ---------- Insufficient sample stays null (BvP) ----------

def test_extract_carries_bvp_null_when_below_floor():
    """BvP payload below MIN_PLATE_APPEARANCES is null at collection; extractor
    carries the null (it does NOT synthesize a value)."""
    db_path = _make_db()
    try:
        conn = sqlite3.connect(db_path)
        _write_snapshot(conn, "7005", "bvp", "2026-07-21", {
            "as_of_utc": "2026-07-21T17:00:00Z",
            "away": None,  # below floor -> null at collection
            "home": {"ops": 0.70, "plate_appearances": 60},
        })
        conn.close()
        rows = build_candidate_dataset(
            db_path,
            game_rows=[{"game_pk": "7005", "date_ymd": "2026-07-21",
                        "as_of_utc": "2026-07-21T17:00:00Z",
                        "first_pitch_utc": "2026-07-21T23:05:00Z"}],
        )
        v = rows["7005"].feature_vector
        assert v["awayLineupBvpOps"] is None
        assert v["homeLineupBvpOps"] == 0.70
    finally:
        os.unlink(db_path)


# ---------- Determinism / hash parity ----------

def test_extract_hash_is_deterministic_across_calls():
    db_path = _make_db()
    try:
        conn = sqlite3.connect(db_path)
        _write_snapshot(conn, "7006", "statcast_team", "2026-07-21", _statcast_payload())
        conn.close()
        game_row = {"game_pk": "7006", "date_ymd": "2026-07-21",
                    "as_of_utc": "2026-07-21T17:00:00Z",
                    "first_pitch_utc": "2026-07-21T23:05:00Z"}
        r1 = build_candidate_dataset(db_path, game_rows=[game_row])["7006"]
        r2 = build_candidate_dataset(db_path, game_rows=[game_row])["7006"]
        assert r1.feature_vector_hash == r2.feature_vector_hash
    finally:
        os.unlink(db_path)


# ---------- Coverage summary ----------

def test_coverage_summary_reports_available_and_missing():
    db_path = _make_db()
    try:
        conn = sqlite3.connect(db_path)
        for gpk in ["8001", "8002"]:
            _write_snapshot(conn, gpk, "statcast_team", "2026-07-21", _statcast_payload())
        _write_snapshot(conn, "8001", "pitch_arsenal", "2026-07-21", _arsenal_payload())
        conn.close()
        rows = build_candidate_dataset(
            db_path,
            game_rows=[
                {"game_pk": gpk, "date_ymd": "2026-07-21",
                 "as_of_utc": "2026-07-21T17:00:00Z",
                 "first_pitch_utc": "2026-07-21T23:05:00Z"}
                for gpk in ["8001", "8002"]
            ],
        )
        summary = candidate_coverage_summary(rows)
        assert summary["row_count"] == 2
        assert summary["families"]["statcast_team"]["available"] == 2
        assert summary["families"]["statcast_team"]["coverage_pct"] == 100.0
        assert summary["families"]["pitch_arsenal"]["available"] == 1
        assert summary["families"]["pitch_arsenal"]["coverage_pct"] == 50.0
    finally:
        os.unlink(db_path)


def test_coverage_summary_empty():
    assert candidate_coverage_summary({}) == {"row_count": 0, "families": {}, "temporal_rejections": {}}

File: tests/test_p4_candidate_features.js
import assert from 'node:assert/strict';
import test from 'node:test';
import { mkdirSync, mkdtempSync, rmSync } from 'node:fs';
import { join, resolve } from 'node:path';
import { tmpdir } from 'node:os';

import { Storage } from '../src/storage.js';
import {
  buildStatcastTeamPayload,
  buildPitchArsenalPayload,
  buildExpectedInningsPayload,
  buildBullpenQualityPayload,
  buildLineupBatterPayload,
  buildBvpPayload,
  buildStarterXeraPayload,
  writeCandidateSnapshots,
  isPromotionSafe,
  CANDIDATE_GROUPS,
  CANDIDATE_FEATURE_SCHEMA_VERSION,
  MIN_BVP_PA
} from '../src/core/candidate_features.js';

function freshStorage() {
  const tempDir = resolve(process.cwd(), `.tmp-p4-tests-${Date.now()}-${Math.random().toString(36).slice(2)}`);
  mkdirSync(tempDir, { recursive: true });
  const statePath = resolve(tempDir, 'state-p4.json');
  return { storage: new Storage(statePath), tempDir };
}

const FIRST_PITCH = '2026-07-21T23:05:00Z';
const AS_OF = '2026-07-21T17:00:00Z'; // pre-pitch
const POST_PITCH = '2026-07-21T23:30:00Z'; // post-pitch

// ---------- Payload builders ----------

test('buildStatcastTeamPayload carries values + sample, nulls absent side', () => {
  const p = buildStatcastTeamPayload({
    away: { xwoba: 0.310, xslg: 0.390, barrelRate: 0.07, hardHitRate: 0.38, samplePa: 120 },
    home: { xwoba: 0.330, xslg: 0.420, barrelRate: 0.09, hardHitRate: 0.41, samplePa: 118 }
  });
  assert.equal(p.away.xwoba, 0.31);
  assert.equal(p.home.barrel_rate, 0.09);
  assert.equal(p.away.sample_pa, 120);
  // absent side -> null
  const p2 = buildStatcastTeamPayload({ away: { xwoba: 0.31, xslg: 0.39, barrelRate: 0.07, hardHitRate: 0.38, samplePa: 5 } });
  assert.equal(p2.home, null);
});

test('buildBullpenQualityPayload computes available innings = 9 - expected starter innings (clamped)', () => {
  const p = buildBullpenQualityPayload({
    away: { qualityScore: 0.6, sampleRelievers: 4, expectedStarterInnings: 6.0 },
    home: { qualityScore: 0.5, sampleRelievers: 3, expectedStarterInnings: 3.0 }
  });
  assert.equal(p.away.available_innings, 3.0); // 9 - 6
  assert.equal(p.home.available_innings, 6.0); // 9 - 3
  // clamp: starter going 0 -> 9 bullpen innings; starter going 12 -> 0
  const clamped = buildBullpenQualityPayload({
    away: { qualityScore: 0.5, sampleRelievers: 3, expectedStarterInnings: 12.0 },
    home: { qualityScore: 0.5, sampleRelievers: 3, expectedStarterInnings: 0.0 }
  });
  assert.equal(clamped.away.available_innings, 0);
  assert.equal(clamped.home.available_innings, 9);
  // missing expected innings -> null available
  const noinn = buildBullpenQualityPayload({
    away: { qualityScore: 0.5, sampleRelievers: 3 }
  });
  assert.equal(noinn.away.available_innings, null);
});

test('buildBvpPayload returns null below MIN_BVP_PA (no guess)', () => {
  const low = buildBvpPayload({ away: { ops: 0.75, plateAppearances: MIN_BVP_PA - 1 } });
  assert.equal(low.away, null);
  const ok = buildBvpPayload({ away: { ops: 0.75, plateAppearances: MIN_BVP_PA } });
  assert.equal(ok.away.ops, 0.75);
  assert.equal(ok.away.plate_appearances, MIN_BVP_PA);
});

test('buildLineupBatterPayload encodes confirmation as 0/1/null', () => {
  const p = buildLineupBatterPayload({
    away: { weightedOps: 0.71, battersCaptured: 9, confirmed: true },
    home: { weightedOps: 0.69, battersCaptured: 9, confirmed: false }
  });
  assert.equal(p.away.confirmed, 1);
  assert.equal(p.home.confirmed, 0);
  const unknown = buildLineupBatterPayload({ away: { weightedOps: 0.71, battersCaptured: 9, confirmed: null } });
  assert.equal(unknown.away.confirmed, null);
});

test('all builders stamp no schema/provenance (orchestrator does that)', () => {
  // builders are pure value mappers; provenance attached at write time.
  const p = buildPitchArsenalPayload({ away: { overallStuffPlus: 105, stuffVsLhh: 103, stuffVsRhh: 107, samplePitches: 60 } });
  assert.equal('as_of_utc' in p, false);
  assert.equal(p.away.overall_stuff_plus, 105);
});

// ---------- Promotion-safety gate ----------

test('isPromotionSafe passes when as_of strictly before first pitch', () => {
  assert.equal(isPromotionSafe(AS_OF, FIRST_PITCH).ok, true);
});

test('isPromotionSafe fails post-pitch, missing, and unparseable', () => {
  assert.equal(isPromotionSafe(POST_PITCH, FIRST_PITCH).ok, false);
  assert.ok(isPromotionSafe(POST_PITCH, FIRST_PITCH).reason.includes('post_pitch'));
  assert.equal(isPromotionSafe(null, FIRST_PITCH).ok, false);
  assert.equal(isPromotionSafe(AS_OF, null).ok, false);
  assert.equal(isPromotionSafe('not-a-date', FIRST_PITCH).ok, false);
});

// ---------- Orchestrator: write-once + temporal rejection ----------

test('writeCandidateSnapshots writes all families when pre-pitch + skips empty', () => {
  const { storage, tempDir } = freshStorage();
  try {
    const payloads = {
      statcast_team: buildStatcastTeamPayload({ away: { xwoba: 0.31, xslg: 0.39, barrelRate: 0.07, hardHitRate: 0.38, samplePa: 5 } }),
      pitch_arsenal: buildPitchArsenalPayload({ away: { overallStuffPlus: 105, stuffVsLhh: 103, stuffVsRhh: 107, samplePitches: 60 } }),
      bvp: buildBvpPayload({ away: { ops: 0.75, plateAppearances: 60 } }),
      // expected_innings + bullpen_quality + lineup_batter + starter_xera omitted -> no_payload
    };
    const results = writeCandidateSnapshots(storage, { gamePk: '7001', dateYmd: '2026-07-21', asOfUtc: AS_OF, firstPitchUtc: FIRST_PITCH }, payloads);
    const byGroup = Object.fromEntries(results.map((r) => [r.group, r]));
    assert.equal(byGroup.statcast_team.written, true);
    assert.equal(byGroup.pitch_arsenal.written, true);
    assert.equal(byGroup.bvp.written, true);
    assert.equal(byGroup.expected_innings.written, false);
    assert.equal(byGroup.expected_innings.reason, 'no_payload');
    // written payloads carry provenance
    const row = storage.getFeatureSnapshot('7001', 'statcast_team');
    assert.equal(row.payload.as_of_utc, AS_OF);
    assert.equal(row.payload.schema_version, CANDIDATE_FEATURE_SCHEMA_VERSION);
  } finally {
    storage.close();
    rmSync(tempDir, { recursive: true, force: true });
  }
});

test('writeCandidateSnapshots is write-once (second call does not overwrite)', () => {
  const { storage, tempDir } = freshStorage();
  try {
    const payloads = {
      statcast_team: buildStatcastTeamPayload({ away: { xwoba: 0.31, xslg: 0.39, barrelRate: 0.07, hardHitRate: 0.38, samplePa: 5 } })
    };
    const ctx = { gamePk: '7002', dateYmd: '2026-07-21', asOfUtc: AS_OF, firstPitchUtc: FIRST_PITCH };
    const r1 = writeCandidateSnapshots(storage, ctx, payloads);
    assert.equal(r1.find((r) => r.group === 'statcast_team').written, true);
    // mutate payload, write again -> should NOT overwrite (write-once).
    const payloads2 = {
      statcast_team: buildStatcastTeamPayload({ away: { xwoba: 0.99, xslg: 0.99, barrelRate: 0.99, hardHitRate: 0.99, samplePa: 99 } })
    };
    const r2 = writeCandidateSnapshots(storage, ctx, payloads2);
    assert.equal(r2.find((r) => r.group === 'statcast_team').written, false);
    assert.equal(r2.find((r) => r.group === 'statcast_team').reason, 'already_captured');
    const row = storage.getFeatureSnapshot('7002', 'statcast_team');
    assert.equal(row.payload.away.xwoba, 0.31); // original preserved
  } finally {
    storage.close();
    rmSync(tempDir, { recursive: true, force: true });
  }
});

test('writeCandidateSnapshots rejects all families when post-pitch (temporal ineligible)', () => {
  const { storage, tempDir } = freshStorage();
  try {
    const payloads = {
      statcast_team: buildStatcastTeamPayload({ away: { xwoba: 0.31, xslg: 0.39, barrelRate: 0.07, hardHitRate: 0.38, samplePa: 5 } })
    };
    const results = writeCandidateSnapshots(storage, { gamePk: '7003', dateYmd: '2026-07-21', asOfUtc: POST_PITCH, firstPitchUtc: FIRST_PITCH }, payloads);
    for (const r of results) {
      assert.equal(r.written, false);
      assert.ok(r.reason.includes('temporal_ineligible'), `group ${r.group}: ${r.reason}`);
    }
    // nothing written
    assert.equal(storage.getFeatureSnapshot('7003', 'statcast_team'), null);
  } finally {
    storage.close();
    rmSync(tempDir, { recursive: true, force: true });
  }
});

test('writeCandidateSnapshots skips all-empty payload (no information)', () => {
  const { storage, tempDir } = freshStorage();
  try {
    // both sides absent -> empty payload
    const payloads = {
      statcast_team: buildStatcastTeamPayload({})
    };
    const results = writeCandidateSnapshots(storage, { gamePk: '7004', dateYmd: '2026-07-21', asOfUtc: AS_OF, firstPitchUtc: FIRST_PITCH }, payloads);
    assert.equal(results.find((r) => r.group === 'statcast_team').reason, 'empty_payload');
    assert.equal(storage.getFeatureSnapshot('7004', 'statcast_team'), null);
  } finally {
    storage.close();
    rmSync(tempDir, { recursive: true, force: true });
  }
});

test('CANDIDATE_GROUPS lists all seven families', () => {
  assert.deepEqual([...CANDIDATE_GROUPS].sort(), [
    'bullpen_quality', 'bvp', 'expected_innings', 'lineup_batter',
    'pitch_arsenal', 'statcast_team', 'starter_xera'
  ].sort());
});
